### PR TITLE
Introduce StrategyRegistry

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Part of #10549 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @akihironitta

### CODE REVIEW

File: pytorch_lightning/strategies/ddp.py [hunk 1]
```diff
@@ -424,8 +424,8 @@ def post_training_step(self):
             self.model.require_backward_grad_sync = True
 
     @classmethod
-    def register_plugins(cls, plugin_registry: Dict) -> None:
-        plugin_registry.register(
+    def register_strategies(cls, strategies_registry: Dict) -> None:
+        strategies_registry.register(
```
Comment (carmocca) [inline pytorch_lightning/strategies/ddp.py:428]: Shouldnt it be "strategy registry"? Not plural

File: pytorch_lightning/strategies/strategies_registry.py [hunk 2]
```diff
@@ -21,46 +21,46 @@
...
-    to the plugins flag to the Trainer.
+    The motivation for having a StrategiesRegistry is to make it convenient
+    for the Users to try different Strategies by passing just strings
+    to the strategy flag to the Trainer.
 
     Example::
 
-        @TrainingTypePluginsRegistry.register("lightning", description="Super fast", a=1, b=True)
+        @StrategiesRegistry.register("lightning", description="Super fast", a=1, b=True)
         class LightningPlugin:
```
Comment (awaelchli) [inline pytorch_lightning/strategies/strategies_registry.py:39]: Could you update this line too :) <3 
```suggestion
        class LightningStrategy:
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/strategies/ddp.py]: Updated!

File: pytorch_lightning/strategies/strategies_registry.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/strategies/strategies_registry.py]: Updated!

File: pytorch_lightning/strategies/ddp_spawn.py [hunk 3]
```diff
@@ -362,8 +362,8 @@ def get_from_queue(self, trainer: "pl.Trainer", queue: "_FakeQueue") -> None:
         trainer.callback_metrics.update(apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x)))
 
     @classmethod
-    def register_plugins(cls, plugin_registry: Dict) -> None:
-        plugin_registry.register(
+    def register_strategies(cls, strategy_registry: Dict) -> None:
+        strategy_registry.register(
             "ddp_spawn_find_unused_parameters_false",
             cls,
             description="DDPSpawn Plugin with `find_unused_parameters` as False",
```
Comment (four4fish) [inline pytorch_lightning/strategies/ddp_spawn.py:369]:
```suggestion
            description="DDPSpawn Strategy with `find_unused_parameters` as False",
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -330,14 +330,14 @@ def handle_given_plugins(self) -> None:
...
+            if isinstance(plug, str) and plug in StrategyRegistry:
                 if strategy is None:
-                    strategy = TrainingTypePluginsRegistry.get(plug)
+                    strategy = StrategyRegistry.get(plug)
                 else:
                     raise MisconfigurationException(
                         "You can only specify one precision and one training type plugin."
                         " Found more than 1 training type plugin:"
-                        f' {TrainingTypePluginsRegistry[plug]["plugin"]} registered to {plug}'
+                        f' {StrategyRegistry[plug]["plugin"]} registered to {plug}'
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:340]:
```suggestion
                        f' {StrategyRegistry[plug]["strategy"]} registered to {plug}'
```

**Review by four4fish — COMMENTED** [review]
LGTM! Thank you ~~
Will we have a seperate PR change `plugin` terms to `strategy`?

**Comment by four4fish** [comment]
The strategyRegister is not user facing right? Will this be breaking change?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/strategies/strategy_registry.py [hunk 5]
```diff
@@ -102,36 +102,36 @@ def get(self, name: str, default: Optional[Any] = None) -> Any:
...
-        return "Registered Plugins: {}".format(", ".join(self.keys()))
+        return "Registered Strategies: {}".format(", ".join(self.keys()))
 
 
-TrainingTypePluginsRegistry = _TrainingTypePluginsRegistry()
+StrategyRegistry = _StrategyRegistry()
 
 
-def is_register_plugins_overridden(plugin: type) -> bool:
+def is_register_strategies_overridden(strategy: type) -> bool:
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/strategy_registry.py:119]: should we make these protected?

File: pytorch_lightning/strategies/strategy_registry.py [hunk 6]
```diff
@@ -102,36 +102,36 @@ def get(self, name: str, default: Optional[Any] = None) -> Any:
...
         super_attr = getattr(previous_super_cls, method_name)
     else:
         return False
 
-    return plugin_attr.__code__ is not super_attr.__code__
+    return strategy_attr.__code__ is not super_attr.__code__
 
 
-def call_training_type_register_plugins(root: Path, base_module: str) -> None:
+def call_register_strategies(root: Path, base_module: str) -> None:
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/strategy_registry.py:133]: same here?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -595,15 +595,14 @@ def root_gpu(self) -> Optional[int]:
 
     @staticmethod
     def _is_plugin_training_type(plugin: Union[str, Strategy]) -> bool:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:597]: this needs update?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 8]
```diff
@@ -595,15 +595,14 @@ def root_gpu(self) -> Optional[int]:
...
-        if isinstance(plugin, str) and (plugin in TrainingTypePluginsRegistry or plugin in list(_StrategyType)):
+        if isinstance(plugin, str) and (plugin in StrategyRegistry or plugin in list(_StrategyType)):
             return True
         return isinstance(plugin, Strategy)
 
     @property
     def is_training_type_in_plugins(self) -> bool:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:603]: this needs update?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 7 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:597]: Accelerator connector cleanup would be addressed in a separate PR as part of #10549

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 8 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:603]: Same here

File: pytorch_lightning/strategies/strategy_registry.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/strategies/strategy_registry.py:119]: I think we can have internal utility functions as unprotected.

File: pytorch_lightning/strategies/strategy_registry.py (hunk 6 above)
Comment (kaushikb11) [reply pytorch_lightning/strategies/strategy_registry.py:133]: Same reason as above.

File: pytorch_lightning/strategies/strategy_registry.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/strategies/strategy_registry.py:119]: As users wouldn't have an use for it, we wouldn't have to worry about users using it.
